[App] Fix env variable login

ethanwharris opened the pull request:
## What does this PR do?

Fixes a bug where login with env variables from the CLI would pass but then not actually create the credentials file.

This is because the CLI login calls `auth = Auth(); auth.clear()` and we were using `__post_init__` to process the environment variables and save the credential file which would then be immediately deleted by the clear call.

No added tests, but added calls to `auth.clear` so the tests reflect the CLI command.

### Does your PR introduce any breaking changes? If yes, please list them.

<!-- FILL IN or None -->

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

<!-- In the CHANGELOG, separate each item in the unreleased section by a blank line to reduce collisions -->

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃


cc @borda

awaelchli reviewed (approved):
@ethanwharris FYI, with the changes in src/lightning_app/utilities/login.py undone, the tests in login.py still pass.

I did:
```py
gh pr checkout 16339
git checkout master src/lightning_app/utilities/login.py
py.test -v  tests/tests_app/utilities/test_login.py
```

ethanwharris commented:
Thanks @awaelchli !!! I've updated the tests to actually check for the credentials file and also refactored the Auth class a bit further to do all authentication in one place. Mind re-reviewing?

awaelchli reviewed (approved):
Looks great!
